[WASM-Function-References] Improve type printing for reference types

takikawa opened the pull request:
#### fd124818fa58bc8e5a9be06c03c1e29ad2564ddd
<pre>
[WASM-Function-References] Improve type printing for reference types
<a href="https://bugs.webkit.org/show_bug.cgi?id=247746">https://bugs.webkit.org/show_bug.cgi?id=247746</a>

Reviewed by Justin Michaud.

Print reftypes when typed funcrefs are enabled in (ref null? kind)
format where `kind` is either a heap type (e.g., `i31`, `struct`)
or a concrete type index (e.g., `&lt;struct:0&gt;`, `&lt;func:3&gt;`).

This printing is only used for validation errors for now, as it
relies on being able to access the ModuleInformation.

* JSTests/wasm/function-references/local_init.js:
(async testLocalInit):
* JSTests/wasm/function-references/ref_types.js:
(async testNonNullExternrefIncompatible):
(async testNonNullFuncrefIncompatible):
* JSTests/wasm/gc/any.js:
(testValidation):
(testNullfuncref):
(testNullexternref):
* JSTests/wasm/gc/arrays.js:
(testArrayDeclaration):
* JSTests/wasm/gc/casts.js:
(testFunctionCasts):
(testEqCasts):
* JSTests/wasm/gc/const-exprs.js:
(async testInvalidConstExprs):
* JSTests/wasm/gc/i31.js:
(testI31Get):
* JSTests/wasm/gc/rec.js:
(testRecDeclaration):
* JSTests/wasm/gc/structs.js:
(testStructDeclaration):
* JSTests/wasm/gc/sub.js:
(testSubDeclaration):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::heapTypeKindAsString):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::validationFail const):
(JSC::Wasm::FunctionParser::validationFailHelper const):
(JSC::Wasm::FunctionParser::typeToStringModuleRelative const):

Canonical link: <a href="https://commits.webkit.org/270988@main">https://commits.webkit.org/270988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55cb9c0d2ca534c5152f373139f31da528211c44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5609 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29204 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24681 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24549 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4432 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23169 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3878 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3956 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29839 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23528 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30164 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26177 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2167 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28078 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5431 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33630 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6485 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4436 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7279 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4340 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

takikawa commented on Source/JavaScriptCore/wasm/WasmParser.h:
```diff
@@ -107,6 +107,7 @@ class Parser {
         using namespace FailureHelper; // See ADL comment in namespace above.
         return UnexpectedResult(makeString("WebAssembly.Module doesn't parse at byte "_s, String::number(m_offset), ": "_s, makeString(args)...));
     }
+
```
whoops this was included by accident, will delete before merge

takikawa replied:
fixed

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -271,10 +271,52 @@ class FunctionParser : public Parser<void>, public FunctionParserTypes<typename
+            // FIXME: use name section if it exists to provide a nicer name.
+            else {
+                const auto& typeDefinition = TypeInformation::get(type.index);
+                if (typeDefinition.expand().is<FunctionSignature>())
```
can we avoid expanding multiple times?

takikawa replied:
This should be fixed now

justinmichaud reviewed (approved):
r=me

takikawa committed:
[WASM-Function-References] Improve type printing for reference types
https://bugs.webkit.org/show_bug.cgi?id=247746

Reviewed by Justin Michaud.

Print reftypes when typed funcrefs are enabled in (ref null? kind)
format where `kind` is either a heap type (e.g., `i31`, `struct`)
or a concrete type index (e.g., `<struct:0>`, `<func:3>`).

This printing is only used for validation errors for now, as it
relies on being able to access the ModuleInformation.

* JSTests/wasm/function-references/local_init.js:
(async testLocalInit):
* JSTests/wasm/function-references/ref_types.js:
(async testNonNullExternrefIncompatible):
(async testNonNullFuncrefIncompatible):
* JSTests/wasm/gc/any.js:
(testValidation):
(testNullfuncref):
(testNullexternref):
* JSTests/wasm/gc/arrays.js:
(testArrayDeclaration):
* JSTests/wasm/gc/casts.js:
(testFunctionCasts):
(testEqCasts):
* JSTests/wasm/gc/const-exprs.js:
(async testInvalidConstExprs):
* JSTests/wasm/gc/i31.js:
(testI31Get):
* JSTests/wasm/gc/rec.js:
(testRecDeclaration):
* JSTests/wasm/gc/structs.js:
(testStructDeclaration):
* JSTests/wasm/gc/sub.js:
(testSubDeclaration):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::heapTypeKindAsString):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::validationFail const):
(JSC::Wasm::FunctionParser::validationFailHelper const):
(JSC::Wasm::FunctionParser::typeToStringModuleRelative const):

Canonical link: https://commits.webkit.org/270988@main
